File: src/SerializableResource.cpp
#include "SerializableResource.h"
#include <godot_cpp/core/class_db.hpp>

Dictionary SerializableResource::_serialize()
{
    Dictionary dict;
    return dict;
}

void SerializableResource::_bind_methods()
{
    ClassDB::add_virtual_method("SerializableResource", MethodInfo(Variant::DICTIONARY, "_serialize"));
    ClassDB::add_virtual_method("SerializableResource", MethodInfo("_deserialize", PropertyInfo(Variant::DICTIONARY, "data")));
}

File: demo/CSharpBindings/SerializableResource.cs
using Godot;
using Godot.Collections;

public abstract partial class SerializableResource : Resource
{
    private Dictionary _serialize()
    {
        Serialize(out Dictionary<string, Variant> dict);
        Variant varDict = dict;

        return varDict.As<Dictionary>();
    }

    private void _deserialize(Dictionary data)
    {
        Variant varDict = data;

        Deserialize(varDict.As<Dictionary<string, Variant>>());
    }

    public abstract void Serialize(out Dictionary<string, Variant> data);
    public abstract void Deserialize(Dictionary<string, Variant> data);
}